OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs

```diff
@@ -0,0 +1,234 @@
+		int scanInterval;
+		int ticks;
+
+		// MCVs that the bot already knows about. Any MCV not on this list needs to be given an order.
+		List<Actor> activeMCVs = new List<Actor>();
+
+		public McvManagerBotModule(Actor self, McvManagerBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+
+			if (world.Type == WorldType.Editor)
+				return;
+
+			unitCannotBeOrdered = a => a.Owner != player || a.IsDead || !a.IsInWorld;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			notifyPositionsUpdated = player.PlayerActor.TraitsImplementing<IBotPositionsUpdated>().ToArray();
+			requestUnitProduction = player.PlayerActor.TraitsImplementing<IBotRequestUnitProduction>().ToArray();
+
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			scanInterval = world.LocalRandom.Next(Info.ScanForNewMcvInterval, Info.ScanForNewMcvInterval * 2);
+		}
+
+		void IBotPositionsUpdated.UpdatedBaseCenter(CPos newLocation)
+		{
+			initialBaseCenter = newLocation;
+		}
+
+		void IBotPositionsUpdated.UpdatedDefenseCenter(CPos newLocation) { }
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			ticks++;
+
+			if (ticks == 1)
+				DeployMcvs(bot, false);
+
+			if (--scanInterval <= 0)
+			{
+				scanInterval = Info.ScanForNewMcvInterval;
+				DeployMcvs(bot, true);
+
+				// No construction yards - Build a new MCV
+				if (ShouldBuildMCV)
+				{
+					var unitBuilder = requestUnitProduction.FirstOrDefault(Exts.IsTraitEnabled);
+					if (unitBuilder != null)
+					{
+						var mcvInfo = AIUtils.GetInfoByCommonName(Info.McvTypes, player);
+						unitBuilder.RequestUnitProduction(bot, mcvInfo.Name);
+					}
+				}
+			}
+		}
+
+		bool ShouldBuildMCV
+		{
+			get
+			{
+				// Only build MCV if we don't already have one in the field.
+				var allowedToBuildMCV = AIUtils.CountActorByCommonName(Info.McvTypes, player) == 0;
+
+				// Require at least one construction yard, unless we have no vehicles factory (can't build it).
+				var hasAdequateContructionYards = AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) > 0 ||
+					AIUtils.CountBuildingByCommonName(Info.McvFactoryTypes, player) == 0;
+
+				return allowedToBuildMCV && !hasAdequateContructionYards;
+			}
+		}
+
+		void DeployMcvs(IBot bot, bool chooseLocation)
+		{
+			activeMCVs.RemoveAll(unitCannotBeOrdered);
+
+			var newMCVs = world.ActorsHavingTrait<IPositionable>()
+				.Where(a => a.Owner == player &&
+					!activeMCVs.Contains(a) &&
+					a.IsIdle &&
+					Info.McvTypes.Contains(a.Info.Name));
+
+			foreach (var a in newMCVs)
+				activeMCVs.Add(a);
+
+			foreach (var mcv in activeMCVs)
+				DeployMcv(bot, mcv, chooseLocation);
+		}
+
+		// Find any MCV and deploy them at a sensible location.
+		void DeployMcv(IBot bot, Actor mcv, bool move)
+		{
+			// Don't try to move and deploy an undeployable actor
+			var transformsInfo = mcv.Info.TraitInfoOrDefault<TransformsInfo>();
```

Comment:
Can we include this in the `newMCVs` filtering instead? This will prevent a bunch of useless references junking up `activeMCVs` and eating perf in mods that require this check.